### PR TITLE
Pod toleration based on CPU and memory

### PR DESCRIPTION
**Pull request by cloudw**
Add toleration for pods requesting significant memory and CPUs. 

TODOs for yunw:
- [x] After merging, backport the same change onto py3.6 version of metaflow (Several AS are staying on old versions to finish up their experiments)
- [x] When both version ready, update each customer project to the newer version. If customers are known to stay on older version, provide them with the version handle.

### CODE REVIEW

**Comment by cloudw** [comment]
Doing a slight refactoring and fixing unit tests.

**Comment by cloudw** [comment]
Slightly refactored the existing affinity and toleration tests for accelerator deco, to also test tolerations for large pod. Moved this test into separate file due to its nature (it's unit test not integration test).

File: metaflow/plugins/kfp/tests/test_pod_toleration_and_affinity.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    assert not has_node_toleration(
+        step_template=step_templates["small_memory_pod"],
+        key="node.kubernetes.io/instance-type",
+        value="r5.12xlarge",
+    )
+    assert has_node_toleration(
+        step_template=step_templates["large_cpu_pod"],
+        key="node.kubernetes.io/instance-type",
+        value="r5.12xlarge",
+    )
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/tests/test_pod_toleration_and_affinity.py]: great test!

File: metaflow/plugins/kfp/kfp.py [hunk 2]
```diff
@@ -564,6 +565,34 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
         cmds.append(" ".join(entrypoint + top_level + step))
         return " && ".join(cmds)
 
+    @staticmethod
+    def _create_resource_based_node_type_toleration(
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:569]: can you please add a comment like
`# TODO(yunw): A Metaflow plugin framework to customize tags, labels, etc.`

Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:569]: also reference `AIP-5264 MutatingWebHook (or Validating) for AIP pod scheduling policies`

File: metaflow/plugins/kfp/kfp.py [hunk 3]
```diff
@@ -643,6 +672,17 @@ def _set_container_resources(
             container_op.add_affinity(affinity)
             container_op.add_toleration(toleration)
 
+        if (
+            not kfp_component.accelerator_decorator
+            and "gpu" not in resource_requirements
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:677]: this should probably be an `else` case for `if kfp_component.accelerator_decorator:`

File: metaflow/plugins/kfp/kfp.py [hunk 4]
```diff
@@ -564,6 +565,34 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
...
+        Resource threshold are lower than machine resource boundary to take overheads into
+        account.
+
+        Toleration allows pods to utilize larger nodes without enforcement.
+        Setting a low threshold for using larger host allow more freedom for scheduler
+        and potentially higher utilization rate.
+        """
+        # No need to validate value - already done by set_<resource>_request above
+
+        if cpu >= 12 or memory >= 24:  # 80% of hard limit - c5.4xlarge: 16 vCPU, 32 GB
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:586]: now: how many resources does the system use? Can we be a bit more aggressive?

```
16-12 = 4 cpus?
32 - 24 = 8 GB?
```

Maybe 
`if cpu >= 14 or memory >= 28:`

Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:586]: Moving the conversation to slack

File: metaflow/plugins/kfp/kfp.py (hunk 2 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:569]: @talebzeghmi we had PR for label and annotation but decided to close it to not leave interface for customer: https://github.com/zillow/metaflow/pull/59. Do you think we should add back these decorator?

File: metaflow/plugins/kfp/kfp.py (hunk 4 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:586]: Updated threshold based on resource deep dive in https://docs.google.com/document/d/1dKkQke78gqjAV2Ww_JnEBipHy9zs2NXCQHailhFiOIs/edit#

Summary of available resource investigation added to comments.

File: metaflow/plugins/kfp/kfp.py [hunk 5]
```diff
@@ -564,6 +565,49 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
...
+        # Memory: 20.24GB = 18.85 GiB = 27.72 GiB (allocatable) - 8.87 GiB (DaemonSet)
+        # CPU: 10.34 vCPU = 15.89 (allocatable) - 5.55 (DaemonSet)
+        # Argo additionally adds a "wait" container to pods per step, taking default resources
+
+        # Threshold should leave significant margin below estimated available resource
+        # Pods will be unschedulable if its resource requirement falls in range
+        #   [available resource, thresholdfor large node)
+
+        # Using 50% of total available resource on c5.4xlarge
+        memory_threshold = 16  # GB
```
Comment (hsezhiyan) [inline metaflow/plugins/kfp/kfp.py:598]: Unless I'm missing something, shouldn't this `memory_threshold` be `10` = `20.24GB / 2 = ~10GB`.

File: metaflow/plugins/kfp/kfp.py [hunk 6]
```diff
@@ -564,6 +565,49 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
...
+        r5.12xlarge: 48 vCPU, 384 GB
+        Resource threshold are lower than machine resource boundary to take overheads into
+        account.
+
+        Toleration allows pods to utilize larger nodes without enforcement.
+        Setting a low threshold for using larger host allow more freedom for scheduler
+        and potentially higher utilization rate.
+
+        cpu: number of vCPU requested. Fractions are allowed.
+        memory: memory requested in GB (not GiB)
```
Comment (hsezhiyan) [inline metaflow/plugins/kfp/kfp.py:586]: Do you mind adding a short description on the cost issue that Rajesh mentioned (that is the reason for this PR)? e.g. something on the lines of `this toleration ensures small pods do not get scheduled on r5.12xlarge nodes and thus do not dramatically increase the cost of cluster usage`..

File: metaflow/plugins/kfp/kfp.py (hunk 5 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:598]: Oh I should make the doc string more clear - by total available I mean 50% of the machine resource. 
16G is already 4 G under the available resource.

I was thinking about this when writing the 50% in docstring:

> Percentage-wise, portion available to users:
> Memory:  `61.65%` of total memory = 20.24 GB /  32.83 GB
> CPU: `64.625%` of total vCPUs10.34 vCPU = 15.89 (allocatable) - 5.55 (daemon set in stage cluster)

Comment (hsezhiyan) [reply metaflow/plugins/kfp/kfp.py:598]: Got it, that makes sense! If possible then, I'd add a comment after `16` with `# = 32.83GB (total available) / 2 = ~16GB`.

File: metaflow/plugins/kfp/kfp.py [hunk 7]
```diff
@@ -564,6 +565,53 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
...
+    @staticmethod
+    def _create_resource_based_node_type_toleration(
+        cpu: float, memory: float
+    ) -> Optional[V1Toleration]:
+        """Allow large enough pod to use higher cost nodes by adding toleration.
+
+        Together with taint added at cluster side,
+        this is a temporary solution to fix "r5.12xlarge host not scaling down" issue,
+        caused by smaller pods keep being scheduled onto large nodes.
+        TODO: Replace using AIP-5264 MutatingWebHook (or Validating) for AIP pod scheduling policies
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:577]: can you also add this todo:

`        # TODO(yunw): A Metaflow plugin framework to customize tags, labels, etc.`

**Review by talebzeghmi — APPROVED** [review]

**Review by hsezhiyan — APPROVED** [review]

File: metaflow/plugins/kfp/kfp.py (hunk 2 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:569]: Same as the other comment below. Will add in separate PR.

**Pull request merged**